terraform/constellation-cluster/main.tf

```diff
@@ -0,0 +1,110 @@
+  provisioner "local-exec" {
+    command = <<EOT
+         ${path.module}/install-constellation.sh
+    EOT
+  }
+  triggers = {
+    always_run = timestamp()
+  }
+}
+
+// terraform_data resource so that it is run only once
+resource "terraform_data" "config_generate" {
+  provisioner "local-exec" {
+    command = <<EOT
+         ./constellation config generate ${var.csp}
+    EOT
+  }
+  depends_on = [
+    null_resource.ensure_cli
+  ]
+}
+
+
+resource "null_resource" "config" {
+  provisioner "local-exec" {
+    command = <<EOT
+      if [ "${var.csp}" = "aws" ]; then
+      yq eval '.provider.aws.region = "${var.aws_config.region}"' -i constellation-conf.yaml
+      yq eval '.provider.aws.zone = "${var.aws_config.zone}"' -i constellation-conf.yaml
+      yq eval '.provider.aws.iamProfileControlPlane = "${var.aws_config.iam_instance_profile_control_plane}"' -i constellation-conf.yaml
+      yq eval '.provider.aws.iamProfileWorkerNodes = "${var.aws_config.iam_instance_profile_worker_nodes}"' -i constellation-conf.yaml
+      fi
+      yq eval '.name = "${var.name}"' -i constellation-conf.yaml
+      if [ "${var.image}" != "" ]; then
+      yq eval '.image = "${var.image}"' -i constellation-conf.yaml
+      fi
+      if [ "${var.kubernetes_version}" != "" ]; then
+      yq eval '.kubernetesVersion = "${var.kubernetes_version}"' -i constellation-conf.yaml
+      fi
+      if [ "${var.microservice_version}" != "" ]; then
+      yq eval '.microserviceVersion = "${var.microservice_version}"' -i constellation-conf.yaml
+      fi
+      ${local.yq_node_groups}
+      ./constellation config fetch-measurements
+    EOT
+  }
+
+  depends_on = [
+    terraform_data.config_generate
+  ]
+
+  triggers = {
+    always_run = timestamp()
+  }
+}
+
+
+resource "null_resource" "infra_state" {
+  provisioner "local-exec" {
+    command = <<EOT
+      yq eval '.infrastructure.uid = "${var.uid}"' -i constellation-state.yaml
+      yq eval '.infrastructure.inClusterEndpoint = "${var.inClusterEndpoint}"' -i constellation-state.yaml
+      yq eval '.infrastructure.clusterEndpoint = "${var.clusterEndpoint}"' -i constellation-state.yaml
+      yq eval '.infrastructure.initSecret = "'"$(echo "${var.initSecretHash}" | tr -d '\n' | hexdump -ve '/1 "%02x"')"'"' -i constellation-state.yaml
+      yq eval '.infrastructure.apiServerCertSANs = ${jsonencode(var.apiServerCertSANs)}' -i constellation-state.yaml
+      yq eval '.infrastructure.name = "${var.name}"' -i constellation-state.yaml
+      yq eval '.infrastructure.ipCidrNode = "${var.ipCidrNode}"' -i constellation-state.yaml
+    EOT
+  }
+  depends_on = [
+    terraform_data.config_generate
+  ]
+  triggers = {
+    always_run = timestamp()
+  }
+}
+
+
+resource "null_resource" "apply" {
+  provisioner "local-exec" {
+    command = "./constellation apply --debug --force --yes --skip-phases infrastructure"
+  } // need force to allow release images while using the dev build
+  // TODO use --debug flag?
```

Comment:
Are the logs of these operations exposed to the user? (or more specifically, can they be not shown to the user explicitly?) If yes, we should enable the debug flag here for sure.

Reply:
I rather meant logging to whatever stream is used for the [TF_LOG logging](https://developer.hashicorp.com/terraform/internals/debugging). This doesn't seem possible on module-level though

Reply:
so should we keep --debug or remove it?

Reply:
better keep it for now than have no logs at all.